File: GA/sequential/src/ga.cpp
#include <vector>
#include <random>
#include <cmath>
#include "ga.hpp"

namespace EA {
  std::vector<std::vector<double>> GA::initialize(int populationSize, int individualSize) {
    std::vector<std::vector<double>> population;
    std::vector<double> individual;
    population.reserve(populationSize);
    individual.reserve(individualSize);
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<> dis(0.0, 1.0);
    for (int i = 0; i < populationSize; i++) {
      individual.clear();
      for (int j = 0; j < individualSize; j++) {
        individual.push_back(dis(gen));
      }
      population.push_back(individual);
    }
    return std::move(population);
  }

  std::vector<std::pair<int, double>> GA::evaluate(std::vector<std::vector<double>>& population, const Bench::BenchMark& benchmark) {
    std::vector<std::pair<int, double>> fitness;
    fitness.reserve(population.size());
    for (int i = 0; i < population.size(); i++) {
      fitness.push_back(std::make_pair(i, benchmark.fitness(population[i])));
    }
    return std::move(fitness);
  }

  // std::vector<std::vector<double>> selection(std::vector<std::vector<double>>&,
  // SelectionHeuristic, double retention, double elite) {

  // }
  // std::vector<double> crossover(CrossoverHeuristic, std::vector<double>&, std::vector<double>&);
  // std::vector<double> mutation(MutationHeuristic, std::vector<double>&, std::vector<double>&);
  std::vector<std::vector<double>> GA::iterate(std::vector<std::vector<double>>& population, double retention, double elite,
    const Heuristcs::SelectionHeuristic& selectionHeuristic, const Heuristcs::CrossoverHeuristic& crossoverHeuristic, const Heuristcs::MutationHeuristic& mutationHeuristic, const Bench::BenchMark& benchmark) {
      std::vector<std::pair<int, double>> fitness = evaluate(population, benchmark);
      std::vector<double> params = {elite, retention};
      std::vector<double> distributionIndex = {5};
      std::vector<std::vector<double>> selectedPopulation = selectionHeuristic.select(population, fitness, params);
      std::vector<std::vector<double>> offsprings;
      for (int i = std::floor(population.size()*elite); i < population.size(); i+=2) {
        offsprings = crossoverHeuristic.crossover(selectedPopulation[i], selectedPopulation[i+1], distributionIndex);
        selectedPopulation[i] = mutationHeuristic.mutate(offsprings[0]);
        selectedPopulation[i+1] = mutationHeuristic.mutate(offsprings[1]);
      }
      return std::move(selectedPopulation);
  }
}

File: GA/haskell/sequential/src/GA.hs
-- | This module implements the core functionality of a Genetic Algorithm (GA).
module GA (initializePopulation, gaIterate) where 

import System.Random (newStdGen, randomRs)
import Data.List.Split (chunksOf)
import Heuristics (simiulatedBinary, gaussianMutate, tournamentSelection)
import Benchmark (sphere)
import Helpers (beta, randomNumbers, normalNoise)

-- | Initializes a population of individuals for the GA.
--
-- Generates a population of 'numIndividual' individuals, each with 'chromosome' genes,
-- where each gene is a random value in the range [-1, 1].
--
-- Arguments:
--   * numIndividual - The number of individuals in the population.
--   * chromosome - The number of genes per individual.
--
-- Returns: An IO action yielding a list of individuals, where each individual is a list of doubles.
initializePopulation :: Int -> Int -> IO ([[Double]])
initializePopulation numIndividual chromosome  = do
  gen <- newStdGen
  let population = take (numIndividual * chromosome) (randomRs (-1, 1) gen :: [Double])
  return $ chunksOf chromosome population

-- | Evaluates the fitness of a population using a given fitness function.
--
-- Applies the fitness function 'f' to each individual in the population.
--
-- Arguments:
--   * population - The list of individuals to evaluate.
--   * f - The fitness function to apply to each individual.
--
-- Returns: A list of fitness values corresponding to the population.
evaluate :: [[Double]] -> ([Double] -> Double) -> [Double]
evaluate population f = map f population

-- | Performs one iteration of the GA, evolving the population toward an optimal solution.
--
-- Iterates up to 10,000 times, using elitism, tournament selection, simulated binary crossover,
-- and Gaussian mutation. Stops at 10,000 iterations and returns the best fitness value.
--
-- Arguments:
--   * iter - The current iteration number (starts at 0).
--   * population - The current population of individuals.
--   * elites - The number of elite individuals to preserve unchanged.
--   * selectionSize - The size of the tournament for selection.
--   * distributionIndex - The distribution index for simulated binary crossover.
--
-- Returns: An IO action yielding the minimum fitness value after iterations complete.
gaIterate :: Int -> [[Double]] -> Int -> Int -> Double -> IO (Double)
gaIterate iter population elites selectionSize distributionIndex
  | iter == 10000 = do
    putStrLn "Iterations completed"
    return $ minimum $ evaluate population sphere
  | otherwise = do
    let populationSize = length population
    beta_ <- beta distributionIndex
    randNS <- randomNumbers (populationSize*selectionSize) (populationSize - 1)
    let selectedPopulation = tournamentSelection population (evaluate population sphere) (randNS) elites selectionSize
    let elitePop = take elites selectedPopulation
    let nonElitePop = drop elites selectedPopulation
    let offspring = concat $ zipWith (\x y -> simiulatedBinary x y beta_) (take (div (populationSize - elites) 2) nonElitePop) (drop (div (populationSize - elites) 2) nonElitePop)
    let offspringSize = length $ offspring!!0
    randNM <- randomNumbers offspringSize 9
    normalN <- normalNoise offspringSize
    let mutated = map (\x -> gaussianMutate x (randNM) (normalN)) offspring
    let newPopulation = elitePop ++ mutated
    gaIterate (iter+1) newPopulation elites selectionSize distributionIndex
